### PR TITLE
Protocol buffers for new compiler API

### PR DESCRIPTION
**Pull request by RobertBColton**
Just testing the waters here by converting the existing Background struct over to proto3. Goal here is to create a more flexible API that doesn't break when resources change. Honestly pretty straight forward, I am committing the codegen from the protobuf compiler to make this easier for newcomers to build it.

Some things that bother me so far: 
* Couldn't exactly find out why the required option is removed from proto3. There are other unexplained differences between proto2 and proto3 as well.
* It looks like Image got included in the Background codegen inline. I compiled this with ``protoc Background.proto --cpp_out=.`` and isn't there a build system like Make for this thing?
* Another particular problem I've noticed is that proto3 apparently will not deserialize unknown fields, which could potentially lead to loss of data if say you opened a new project under an older build of the compiler API I am now creating?

https://developers.google.com/protocol-buffers/docs/proto3#unknowns

> For most Google protocol buffers implementations, unknown fields are not accessible in proto3 via the corresponding proto runtimes, and are dropped and forgotten at deserialization time.

Let me know what you think so far.

### CODE REVIEW

File: CompilerSource/assend/Background.proto [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+import "Image.proto";
+
+message Background {
+  string name = 1;
+  int32 id = 2;
+
+  bool transparent = 3;
+  bool smoothEdges = 4;
+  bool preload = 5;
+  bool useAsTileset = 6;
```
Comment (JoshDreamland) [inline CompilerSource/assend/Background.proto]: What we often do is extract a separate message for related concepts. I didn't expect you to start with backgrounds,  or I'd have synced with you on that, first. My intention is for these field names to mirror EDL constructs—we should keep track of that and document it as we design EGM.

File: CompilerSource/assend/Background.proto [hunk 2]
```diff
@@ -0,0 +1,21 @@
+syntax = "proto3";
+
+import "Image.proto";
+
+message Background {
+  string name = 1;
+  int32 id = 2;
+
+  bool transparent = 3;
+  bool smoothEdges = 4;
```
Comment (JoshDreamland) [inline CompilerSource/assend/Background.proto]: Style: snake case for these (eg, `smooth_edges`).

**Comment by RobertBColton** [comment]
> Right. I advise sticking to Proto2 for this reason, and others (eg, has_integer_field).
How certain are you that Google is going to keep Proto2 around long enough though?

That alone is making me want to start considering alternatives already.

Actually... what the fuck is this?
https://google.github.io/flatbuffers/

Looks like Google invented some buffers just for games.

**Comment by RobertBColton** [comment]
Closing this proposal because we've discovered in #1116 that FlatBuffers is more appropriate for our use case.

**Pull request closed without merging**